py-development.Rmd

```diff
@@ -0,0 +1,520 @@
+    for number in list_of_numbers:
+        number_sum += number
+    return number_sum
+
+sum_all_numbers([1, 2, 3,])
+```
+
+A function can also return multiple outputs, e.g. we can return the number of
+elements in addition to their sum:
+
+```{python}
+def sum_and_len_all_numbers(list_of_numbers):
+    number_sum = 0
+    number_len = 0
+    for number in list_of_numbers:
+        number_sum += number
+        number_len += 1
+    return number_sum, number_len
+```
+
+To capture the output, we can either assign to a single name, a tuple, or
+assign both at the same time to different variables.
+
+```{python}
+sum_and_len_of_numbers = sum_and_len_all_numbers([1, 2, 3])
+sum_and_len_of_numbers
+```
+
+```{python}
+sum_of_numbers, len_of_numbers = sum_and_len_all_numbers([1, 2, 3])
+sum_of_numbers
+```
+
+```{python}
+len_of_numbers
+```
+
+> Note: When we defined the function `sum_two_numbers()`, we referred to `num1`
+> and `num2` as *parameters*, while we refer to the numbers we pass to the
+> function call (`2` and `5` above) as *arguments*. Although this might sound
+> confusing at first, it is a standard followed by many programming languages
+> so it is useful to get accustomed to this terminology.
+
+#### Function composition {py-function-composition}
+
+Combining functions is referred to as function composition. This practice
+allows us to write functions that perform one specific task and then combine
+them for more complicated tasks, which makes code more readable and easier to
+debug. By composing a function from the built-in `len()` and `sum()` functions,
+we can create a more succinct and easier to read version of our previous
+function `sum_and_len_all_numbers()`.
+
+```{python}
+def sum_and_len(list_of_numbers):
+    return sum(list_of_numbers), len(list_of_numbers)
+```
+
+#### Defining default values {py-defining-default-values}
+
+Up until now, our function calls have included enough arguments to assign one
+to each of the function parameters. This can get tedious in functions with many
+parameters and it is often beneficial to define default values for several of
+the function parameters. As an example, we can modify `sum_two_numbers()` to
+optionally return the two input arguments.
+
+```{python}
+def sum_two_numbers(num1, num2, return_input=False):
+    if return_input:
+        return num1, num2, num1 + num2
+    else:
+        return num1 + num2
+```
+
+By default the function will work just as previously.
+
+```{python}
+sum_two_numbers(2, 5)
+```
+
+But we now also have the option to return the input numbers.
+
+```{python}
+sum_two_numbers(2, 5, return_input=True)
+```
+
+Since the arguments are given in the same order as in the function definition,
+we could have left out the `return_input=` part and just written `True` in the
+third position. If we want to supply the arguments out of order, we need to
+specify the parameter name of each argument.
+
+#### Function documentation {py-function-documentation}
```

Comment:
I am not including the Python type annotations. I guess that would be best practice since it allows IDEs to autodetect easier, but I find it quite messy with default arguments (and I believe many IDEs can read numpy doscstring for types).